### ISSUE TITLE
Track whether DataFrame partitions are sorted by index, even if divisions are unknown?

### ISSUE THREAD

**Issue by gjoseph92** [dask/dask#9425]
I imagine that operations like these are somewhat common:
```python
merged = dd.merge(ddf1, ddf2, on=["a", "b"])
result = merged.groupby(["a", "b"]).apply(custom_func)
```

```python
ab = dd.merge(ddf1, ddf2, on=["a", "b"])
cd = dd.merge(ddf3, ddf4, on=["a", "b"])
final = dd.merge(ab, cd)
```
In these cases, there's a much faster path we could take in `groupby` and `merge` if we knew that the input was already sorted by `["a", "b"]`—even if we didn't know its divisions.

We just need to know that the partitions are already sorted along the index/columns in question—that is, any given value of the group will only be found in exactly one partition. Groupby could become just a `map_partitions` in that case https://github.com/dask/dask/issues/8361, and [`merge` already is internally](https://github.com/dask/dask/blob/57ff5c2e9c550e3c76592b0affa0d45e88375e12/dask/dataframe/multi.py#L381-L390).

What if we had another parameter like `divisions`, which tracked which column(s)/index a DataFrame was sorted by, if any?
* If `df.known_divisions`, then `df.partitioned_by` would always be known
* If `df.partitioned_by`, `df.divisions` might still be unknown. This would be the case after any shuffle-based operation, namely a merge.

Then methods like `groupby` and `merge` could take advantage of this information to offer embarrassingly-parallel paths in these cases.

I know that maintaining divisions for multiple columns has been discussed before, but this seems like a much easier change to make initially, since we wouldn't be changing the meaning of current `divisions`. We could add the field, then add support for it in other methods gradually.

cc @ian-r-rose @jsignell @rjzamora @jrbourbeau

Very related to:
* https://github.com/dask/dask/issues/6246 (somewhat the same as this)
* https://github.com/dask/dask/issues/1493

**Comment by ian-r-rose** [comment]
Another use-case for something like this would be reading in hive-partitioned parquet datasets. That partitioning scheme doesn't fit in well with dask's `divisions` tracking, and it makes certain very obvious operations more difficult than they should be (including merges, groupbys, filters).

Also xref https://github.com/dask/dask/issues/8980

**Comment by jsignell** [comment]
This is another example of per-partition metadata (like length) that if-tracked could provide useful fast paths. This idea has been discussed before and kind of fizzled out in high-level expressions discussions. Maybe the time is ripe to find a middle ground in between divisions and a full high level expression system?

**Comment by gjoseph92** [comment]
I don't think this is even per-partition though—it would be metadata about all the partitions, overall. So hopefully much simpler than that, and wouldn't need to involve high-level expressions at all. I'm imagining:

```python
class DataFrame:
    @property
    def partitioned_by(self) -> tuple[str, ...] | None:
        """Column name(s) the DataFrame is partitioned by

        If `partitioned_by` is not None, then each unique value of those
        columns will be found in exactly one partition.
        
        The "value" is considered as the tuple of all those columns.
        That is, if `partitioned_by` is ``('a', 'b')``, then each unique
        combination of ``a`` and ``b`` will be found in one partition. But
        unique values of ``a`` or ``b``, on their own, may be found in multiple
        partitions. In other words, `partitioned_by` is not hierarchical.

        If `known_divisions` is True, `partitioned_by` will be ``["_index"]``.
        """
        return self._partitioned_by
```

After writing this out, I realize why doing this through an expression system would be a bit nicer. Because it's not hierarchical, I think tracking this would only be useful for a groupby or merge subsequent to a shuffle, on the exact same columns. So it's maybe an overly niche thing to track broadly.

But there are also datasets that I'm sure are partitioned by multiple columns hierarchically (and someday, maybe multi-indexes), which would probably also be useful to track. But since our shuffle is non-hierarchical, we'd need some way of representing both cases in order to make this more generic. You could imagine `partitioned_by` being a list, tuple, or set to represent the different possibilities:
1. list: hierarchical
2. tuple: non-hierarchical
3. set: each column is, independently, guaranteed to be partitioned (I'm not sure if this is even possible without also being hierarchical, and probably not at all common; I just made it up?)

**Comment by jorloplaz** [comment]
Sounds reasonable, but we should be careful with some cases:

> If `known_divisions` is True, `partitioned_by` will be ``["_index"]``.

It's not so simple. In theory the frame should always be sorted by its index (even if we don't know its `divisions`), but there are some cases where this doesn't hold. For example, when you do `ddf = dd.from_pandas(df, sort=False)` and `df` already has an index which is NOT sorted, this index is passed on, and the result is that `ddf.known_divisions = False`, but we cannot say either that `partitioned_by` is the index (because that index is still unsorted). 

I guess the same can happen when reading from some files? The current situation is that you can only be sure that the index is sorted if `ddf.known_divisions is True`.

Moreover, I'd return `["_index"]` only when `self.index.name is None`, because that's the case when the index isn't a column of the original dataframe. When it is indeed a column of the data, I'd return `self.index.name`, because the idea is "OK, this frame is sorted by this column". The fact that the column is also acting as the index is a different thing.

So I propose the following as return:
- `None`: no idea about sortedness, not even about index sortedness.
- `["_index"]`: frame sorted on its anonymous index; rest of columns may be sorted but we don't know.
- `[col_index]`: frame sorted on the column acting current as index; rest of columns may be sorted but we don't know.
- `["_index", "col1", ...]`: frame sorted on its anonymous index, plus some additional columns which are still columns.
- `["col_index", "col1", ...]`: frame sorted on the column acting current as index, plus some additional columns which are still columns.

**Comment by rjzamora** [comment]
> Maybe the time is ripe to find a middle ground in between divisions and a full high level expression system?

The HLG/HLE concern was that the existing practice of defining and tracking `meta` and `divisions` outside of the HLG/HLE was problematic for graph optimization and regeneration. Therefore, it seemed reasonable to wait for `meta` and `divisions` to be moved into the HLG/HLE before redesigning them.  I still think we need to **move** `meta` and `divisions` into `HighLevelGraph`, and I will try to write up a clear proposal for this in #9159. However, I don’t think it is very important that we do that first.  That said, I think the “happy middle ground” would be to introduce a `CollectionMetadata` class where things like `meta`, `divisions`, and `partitioned_by` (and maybe even *custom* metadata) can all be managed in one place. For example:


```python
class DataFrame:

    def __init__(self, dask, name, meta, divisions, **metadata_kwargs):
        self.name = name
        self.dask = dask
        self.collection_metadata = DataFrameCollectionMetadata(
            meta=meta,  # Move meta into self.collection_metadata
            divisions=divisions,  # Move divisions into self.collection_metadata
            **metadata_kwargs,
        )

    @property
    def meta(self):
        return self.collection_metadata.meta

    @property
    def divisions(self):
        return self.collection_metadata.divisions
        
class CollectionMetadata:
    # Any properties shared by DataFrame/Array/Bag?
    pass

class DataFrameCollectionMetadata(CollectionMetadata):

    def __init__(
        self,
        meta,
        divisions,
        partitioned_by=None,
        partition_lens=None,
        **custom_metadata,
    ):
        # Traditional metadata
        self._meta = meta
        self._divisions = divisions
        # New and/or custom metadata
        self._partitioned_by = partitioned_by
        self._partition_lens = partition_lens
        self._custom_metadata = custom_metadata
        
    @property
    def meta(self):
        return self.meta

    @property
    def divisions(self):
        return self.divisions
        
    @property
    def partitioned_by(self):
        return self._partitioned_by
        
    @property
    def partition_lens(self):
        return self._partition_lens
        
    def custom_metadata(self, key):
        return self._custom_metadata[key]
```

In the long run, I’d like both collection metadata and graph materialization logic to live in `HighLevelGraph`. Introducing a `CollectionMetadataClass` (or something like it) should make it a bit easier to accomplish this without needing to change how collections are constructed throughout the entire code base:

```python
class DataFrame:

    def __init__(self, dask, name, meta, divisions, **metadata_kwargs):
        self.dask = dask
        self.dask.validate_name(name)
        self.dask.set_collection_metadata(
            DataFrameCollectionMetadata(
                meta=meta,
                divisions=divisions,
                **metadata_kwargs,
            )
        )

    @property
    def name(self):
        return self.dask.name

    @property
    def collection_metadata(self):
        return self.dask.collection_metadata

    @property
    def meta(self):
        return self.collection_metadata.meta

    @property
    def divisions(self):
        return self.collection_metadata.divisions
```

(Note that this second step requires some HLG simplification and redesign work, but the first part could be done right away)

**Comment by fjetter** [comment]
I believe this issue is mixing up a couple of different concepts.

##  1. Unique partition mapping / primary key

I believe the property this issue is talking about is whether or not a column can be considered a _primary_ partition _key_ (c.f. primary indices in databases), i.e. there is a bijective mapping between value and partition, e.g.

| Partition | value |
|-----------|-------|
| 1         | 4     |
| 1         | 10    |
| 1         | 4     |
| 2         | 6     |
| 3         | 1     |

even if we didn't know the actual values, or rather upper and lower bounds, knowing that there is a unique mapping, i.e. a given value only occurs in a single partition, allows us to perform a trivial aca groupby.

Example: result of a shuffle

## 2. Primary key with different mapping

Knowing that there exists a primary key is not sufficient if we want to compare two different tables. Let's say the above table is TableA and the following is TableB for which `value` also is a primary key


| Partition | value |
|-----------|-------|
| 1         | 4     |
| 1         | 4     |
| 2         | 10    |
| 2         | 6     |
| 3         | 1     |

However, as you will notice, the third row has a different partition<->value mapping, i.e. if we wanted to join those two tables we would still not be able to do this trivially

Example: loading a hive-like parquet dataset

## 3. Sorted (which one?)

I think there are a couple of different definitions possible about what it means to have a sorted dataframe. For example, the following could be considered sorted but it still doesn't offer any useful property to pick a fastpath w/out knowing upper/lower boundaries.

| Partition | value |
|-----------|-------|
| 1         | 1     |
| 2         | 4     |
| 3         | 4     |
| 3         | 6     |
| 3         | 10    |

"sorted partitions" otoh could simply mean (i.e. partition internal sorting)

| Partition | value |
|-----------|-------|
| 1         | 4     |
| 1         | 4     |
| 1         | 10    |
| 2         | 6     |
| 3         | 1     |

I think "sorting" itself is not that useful unless we have neither property 1.) nor `known_divisions` (even if they were overlapping).

---

The reason why I'm writing this down in this details is because

- the title of the issue talks about sorted partitions. I don't think we care for any kind of sorting here but rather if a column (or weaker, a tuples of columns) is a primary key.
- Simply knowing that there is a unique mapping as proposed in https://github.com/dask/dask/issues/9425#issuecomment-1227820265 is not necessarily sufficient to perform efficient merges. We would also need to know that the mapping originates from the same function, i.e. merging a shuffled DF to a hive-like partition DF as suggested in https://github.com/dask/dask/issues/9425#issuecomment-1226511441 is not necessarily possible. However, merging two DFs with same Npartitions that passed through a shuffle (i.e. the same hash/modulo function creates a unique mapping) is possible.

This would render a `partition_by` property definitely more complex since we'd need to track/know how this partitioning was created or else we will only be able to benefit from this knowledge in single-df operations.

**Comment by rjzamora** [comment]
>Simply knowing that there is a unique mapping as proposed in https://github.com/dask/dask/issues/9425#issuecomment-1227820265 is not necessarily sufficient to perform efficient merges. We would also need to know that the mapping originates from the same function

That is exactly right - In https://github.com/dask/dask/pull/9473, I found that `partitioned_by` needed to return **how** a collection was partitioned by a specific set of columns rather than returning a simple True/False. For example, we can only avoid a shuffle if both sides of a merge are shuffled by the same hash function and divided into the same number of partitions.

**Comment by gjoseph92** [comment]
Agreed. To get a fastpath with merges, you'd need to know they'd used the same hash function. When I wrote this, I was thinking of `shuffle` as the only operation that would set `partitioned_by`, so you could expect the same hash function (though not if you used different shuffle methods). If you open this up to IO functions being able to set it too, then you definitely need more information.

For a groupby, though, you don't need any extra information. By knowing that all values from each group must be in exactly one partition, you can always take the fastpath.

The main point here was "shuffles are expensive; once you've done one, how can we get the most value out of it?" It still seems like setting `partitioned_by` would be a pretty cheap way to automatically take some fastpaths when they're available.

Hive-partitioned datasets, or hierarchical partitioning, seem like they should be out of scope. No IO function would set `partitioned_by`.  It would only be set by `shuffle`, to a tuple of `(columns_tuple, shuffle_method)`. Then:
- `groupby` would have a fastpath if this is set at all to improve the groupby-post-shuffle case (the groupby-on-a-partitoned-dataset case already needs to be handled via divisions: https://github.com/dask/dask/issues/8361)
- `merge` would have a fastpath iff `partitioned_by` was equal for both inputs, and they had the same number of partitions
